### PR TITLE
Cythonize call loop

### PR DESCRIPTION
**Pull request by tgoodlet**
Consider this a working initial draft to begin our foray into `cython`-land.
The benchmark suite shows a x2 speedup with this cythonized version of `_multicall` 👍

### CODE REVIEW

**Comment by tgoodlet** [comment]
Apparently `setup_requires` can be used according to this [SO answer](https://stackoverflow.com/a/37471428) if we want to always rely on `cython` as a dep. Looking for opinions here.

**Comment by RonnyPfannschmidt** [comment]
please dont commit the c sources to the git repo, also not exactly what i had in mind for cythonizing :(

**Comment by tgoodlet** [comment]
>please dont commit the c sources to the git repo

Done.

>also not exactly what i had in mind for cythonizing :(

We can always make it better 👍. 
I really just wanted to do this to see how much of a speedup we'd get.

**Comment by RonnyPfannschmidt** [comment]
the biggest problem i see is we broke sane pypy support - c extensions are just not the way to go)

File: tox.ini [hunk 1]
```diff
@@ -9,12 +9,14 @@ deps=
   pytestrelease: pytest
   pytestmaster: git+https://github.com/pytest-dev/pytest.git@master
   pytestfeatures: git+https://github.com/pytest-dev/pytest.git@features
+  cython
 
 [testenv:benchmark]
-commands=py.test {posargs:testing/benchmark.py}
+commands=py.test testing/benchmark.py {posargs}
```
Comment (RonnyPfannschmidt) [inline tox.ini]: it looks like this invocation got accidentally plucked incorrectly

Comment (tgoodlet) [reply tox.ini]: I actually did this on purpose so I could pass additional args without having to pass `testing/benchmark` every time. Should I put it back?

**Comment by tgoodlet** [comment]
>the biggest problem i see is we broke sane pypy support - c extensions are just not the way to go)

So what did you mean by *cythonize* then in our previous discussions?

Also, with regard to `pypy` I agree it seems `cython` [doesn't have great support](http://docs.cython.org/en/latest/src/userguide/pypy.html).
I can keep the original pure python implementation as well and then only import the `cythonized` module when running on `cpython`?

**Comment by RonnyPfannschmidt** [comment]
i recall that cython had a mode where the code was pure python * a separate file with annotation

**Comment by nicoddemus** [comment]
I'm not sure the speed up is worth it; I have the feeling that the bottleneck in applications using `pluggy` is in the actual implementation of the hooks, not the time spent in the multi-call mechanism.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus the speed of the fixture mechanism and the hook calling mechanism is actually critical in some bits

the switch from kwargs to positional args already did a lot to safe time, and i suspect there is more to gain

i recall one of the reasons why pytest doesn't integrate xunit with fixtures is, that the fixtures where too expensive

**Comment by nicoddemus** [comment]
> i recall one of the reasons why pytest doesn't integrate xunit with fixtures is, that the fixtures where too expensive

Probably because the initial implementation always added a `setup/teardown` fixture for all levels (function/class/module) regardless if there were an implementation of `setUp/tearDown` in the `unittest` class; this could be circumvented by only creating the autouse-fixtures when there is a respective xUnit method.

But I digress, if indeed this is a bottleneck let's see if we can squeeze some more performance from it.  👍

**Comment by tgoodlet** [comment]
>i recall that cython had a mode where the code was pure python

Yes there is [pure python mode](http://docs.cython.org/en/latest/src/tutorial/pure.html) though speed gains are much more negligible.

**Comment by hpk42** [comment]
Is there any project where calling hooks is shown to be a speed issue? 
I am using pluggy in 5 different projects and there is not the slightest problem. 
Last time i profiled with pytest the multicall-machinery did not show up in any significant way. 
So i suggest to close this issue and not bother for optimizing unless there is clear evidence it's needed (and even then, there might be easy wins without going full cython).

**Comment by tgoodlet** [comment]
@hpk42 I totally agree. Forgive my premature cythonization ;)

**Pull request closed without merging**